internal/server/receiver_handlers.go

```diff
@@ -70,65 +70,72 @@ func IndexReceiverWebhookPath(o client.Object) []string {
 	return nil
 }
 
-func (s *ReceiverServer) handlePayload() func(w http.ResponseWriter, r *http.Request) {
-	return func(w http.ResponseWriter, r *http.Request) {
-		ctx := context.Background()
-		digest := url.PathEscape(strings.TrimPrefix(r.RequestURI, apiv1.ReceiverWebhookPath))
+func (s *ReceiverServer) handlePayload(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	digest := url.PathEscape(strings.TrimPrefix(r.RequestURI, apiv1.ReceiverWebhookPath))
 
-		s.logger.Info(fmt.Sprintf("handling request: %s", digest))
+	s.logger.Info(fmt.Sprintf("handling request: %s", digest))
 
-		var allReceivers apiv1.ReceiverList
-		err := s.kubeClient.List(ctx, &allReceivers, client.MatchingFields{
-			WebhookPathIndexKey: r.RequestURI,
-		}, client.Limit(1))
-		if err != nil {
-			s.logger.Error(err, "unable to list receivers")
-			w.WriteHeader(http.StatusInternalServerError)
-			return
-		}
+	var allReceivers apiv1.ReceiverList
+	err := s.kubeClient.List(ctx, &allReceivers, client.MatchingFields{
+		WebhookPathIndexKey: r.RequestURI,
+	}, client.Limit(1))
+	if err != nil {
+		s.logger.Error(err, "unable to list receivers")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
 
-		if len(allReceivers.Items) == 0 {
-			w.WriteHeader(http.StatusNotFound)
-			return
-		}
+	if len(allReceivers.Items) == 0 {
+		w.WriteHeader(http.StatusNotFound)
+		return
+	}
 
-		receiver := allReceivers.Items[0]
-		logger := s.logger.WithValues(
-			"reconciler kind", apiv1.ReceiverKind,
-			"name", receiver.Name,
-			"namespace", receiver.Namespace)
+	receiver := allReceivers.Items[0]
+	logger := s.logger.WithValues(
+		"reconciler kind", apiv1.ReceiverKind,
+		"name", receiver.Name,
+		"namespace", receiver.Namespace)
 
-		if receiver.Spec.Suspend || !conditions.IsReady(&receiver) {
-			err := errors.New("unable to process request")
-			if receiver.Spec.Suspend {
-				logger.Error(err, "receiver is suspended")
-			} else {
-				logger.Error(err, "receiver is not ready")
-			}
-			w.WriteHeader(http.StatusServiceUnavailable)
-			return
+	if receiver.Spec.Suspend || !conditions.IsReady(&receiver) {
+		err := errors.New("unable to process request")
+		if receiver.Spec.Suspend {
+			logger.Error(err, "receiver is suspended")
+		} else {
+			logger.Error(err, "receiver is not ready")
 		}
+		w.WriteHeader(http.StatusServiceUnavailable)
+		return
+	}
 
-		if err := s.validate(ctx, receiver, r); err != nil {
-			logger.Error(err, "unable to validate payload")
-			w.WriteHeader(http.StatusBadRequest)
-			return
-		}
+	if err := s.validate(ctx, receiver, r); err != nil {
```

Comment:
As mentioned in the first PR, this line reads the request payload depending on the type of the Receiver. If that happens, `newCELEvaluator()` further below will not be able to read the payload again. My suggestion is that we read the payload entirely into memory at some point earlier than both function calls and pass the payload to both.

Reply:
Ahhh...yes, I forgot about that, I guess I'm not triggering it because of the specific receiver type I'm using.

I will add a test and see what we can do about fixing it 😄

Reply:
Also I suggest reading the payload only after the Receiver object has been successfully fetched/identified to reduce the surface for DDoSing the server.

Reply:
Sorry, looks like there's nothing to fix here! All the places reading the payload inside validate() are adding it back with `r.Body = io.NopCloser(bytes.NewReader(b))`. Sorry about the confusion!